Detectors/TOF/reconstruction/src/Decoder.cxx

```diff
@@ -0,0 +1,200 @@
+#include "TOFReconstruction/Decoder.h"
+#include <iostream>
+#include <chrono>
+#include <algorithm>
+#include "CommonConstants/LHCConstants.h"
+#define VERBOSE
+
+namespace o2
+{
+namespace tof
+{
+namespace compressed
+{
+
+bool Decoder::open(std::string name)
+{
+  if (mFile.is_open()) {
+    std::cout << "Warning: a file was already open, closing" << std::endl;
+    mFile.close();
+  }
+  mFile.open(name.c_str(), std::fstream::in | std::fstream::binary);
+  if (!mFile.is_open()) {
+    std::cerr << "Cannot open " << name << std::endl;
+    return true;
+  }
+
+  if (mBuffer)
+    delete[] mBuffer;
+
+  mFile.seekg(0, mFile.end);
+  mSize = mFile.tellg();
+  mFile.seekg(0);
+
+  mBuffer = new char[mSize];
+
+  // read content of infile
+  mFile.read(mBuffer, mSize);
+  mUnion = reinterpret_cast<Union_t*>(mBuffer);
+  mUnionEnd = reinterpret_cast<Union_t*>(mBuffer + mSize - 1);
+  close();
+
+  return false;
+}
+
+bool Decoder::close()
+{
+  if (mFile.is_open())
+    mFile.close();
+  return false;
+}
+
+void Decoder::readTRM(std::vector<Digit>* digits, int iddl, int orbit, int bunchid)
+{
+  if (mVerbose)
+    printTRMInfo();
+  int nhits = mUnion->frameHeader.numberOfHits;
+  int time_ext = mUnion->frameHeader.frameID << 13;
+  int itrm = mUnion->frameHeader.trmID;
+  mUnion++;
+
+  // read hits
+  Int_t channel, echannel;
+  Int_t tdc;
+  Int_t tot;
+  Int_t bc;
+  Int_t time;
+
+  Int_t digitInfo[4];
+
+  for (int i = 0; i < nhits; i++) {
+    FromRawHit2Digit(iddl, itrm, mUnion->packedHit.tdcID, mUnion->packedHit.chain, mUnion->packedHit.channel, orbit, bunchid, time_ext + mUnion->packedHit.time, mUnion->packedHit.tot, digitInfo);
+
+    if (mVerbose)
+      printHitInfo();
+    digits->emplace_back(digitInfo[0], digitInfo[1], digitInfo[2], digitInfo[3]);
+    mUnion++;
+  }
+}
+
+void Decoder::FromRawHit2Digit(int iddl, int itrm, int itdc, int ichain, int channel, int orbit, int bunchid, int tdc, int tot, int* digitInfo)
+{
+  // convert raw info in digit info (channel, tdc, tot, bc)
+  // tdc = packetHit.time + (frameHeader.frameID << 13)
+  int echannel = Geo::getECHFromIndexes(iddl, itrm, ichain, itdc, channel);
+  digitInfo[0] = Geo::getCHFromECH(echannel);
+  digitInfo[2] = tot;
+
+  digitInfo[3] = int(orbit * o2::tof::Geo::BC_IN_ORBIT);
```

Comment:
orbit will span full uint range, here you multiply it by 3564 and assign to int, this will lead to overflow.

Reply:
Then We have to move
 Int_t mBC;               ///< Bunch Crossing
to Long in the Digit definition.
Do you agree?

Reply:
If I got it correctly, you store in the digit as an orbit just the orbit within the TF rather that that within the run. You should know better what the raw data will provide. If it is the full orbit, you will need either to use extra space in every digit or to subtract from the full orbit the orbit of the TF start and store 0:255 value. In the latter case you will need also to store as a TF parameter this TF start orbit.

Reply:
I need to check what will be provided in the real data. I will discuss internally our group to see what is the better strategy and then implement the better solution. This is still an open discussion